rfc/0010-rancher-integration-policies.md

```diff
@@ -0,0 +1,162 @@
+is `kubewarden`, and/or the URL starts with `https://github.com/kubewarden/`.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+## For policies owned by the Kubewarden team
+
+Needing new Helm charts to ship policies owned by the Kubewarden team.
+
+A pull request bot needs to be created to make the submissions to
+https://github.com/rancher/charts, and we will need external approvals for
+merging.
+
+## For 3rd party policies
+
+Using the provided Artifact Hub API, given that it doesn't have pagination and
+one needs to query for each package independently, will be inefficient. A cache
+could be implemented.
+
+Obtaining the policies via the API is still missing:
+- Metadata not present in `policy/artifacthub-pkg.yml` yet present in
+  `policy/metadata.yml`, namely the array of `rules`.
+- `question.yml` information. Not a problem per se, as 3rd party policies would
+  not need Rancher integration.
+
+Airgap installations wouldn't be possible.
+
+# Alternatives
+[alternatives]: #alternatives
+
+## A. Using the metadata stored in the custom section of Wasm modules
+
+It could be possible to save the `questions.yml` metadata in the metadata
+annotations of `annotated-policy.wasm` themselves (the custom section of the
+Wasm module containing raw data, for us with the name
+`KUBEWARDEN_CUSTOM_SECTION_METADATA`).
+
+With this, annotated policies will contain almost all metadata needed.
+Information about if the policy is signed couldn't be included there by
+definition.
+
+Then, Rancher Explorer frontend could make use of a client for OCI registries
+(such as [oci-registry-js](https://www.npmjs.com/package/oci-registry-js) or a
+wasm-compiled Rust library for example), to pull and cache each Wasm module.
+With [`WebAssembly.Module.customSections()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WebAssembly/Module/customSections)
+it would be possible to extract all the metadata from the annotations of the
+policy.
+
+This would mean to pull hundreds of megabytes, as it will download the Wasm
+module of the policies, regardless of pagination and caching.
+
+## B. Storing policy metadata in the OCI registry
+
+Same as alternative (A), but storing the policies' metadata in the OCI registry
+either under:
+- A new layer, with `config.mediaType` [creating an artifact type
+`application/vnd.oci.yaml-sample.config.v3+yaml`](https://github.com/opencontainers/artifacts/blob/main/artifact-authors.md#defining-a-unique-artifact-type).
+  This means pulling layers, and possibly uncompressing them.
+- Or in `manifest.config` (https://github.com/opencontainers/artifacts/blob/main/artifact-authors.md#optional-defining-config-schema)
+
+It would then be retrieved with an OCI registry client. This would be more
+efficient than pulling full Wasm modules for all policies as in (A).
+
+Airgap installations would consist on mirroring the registry.
+
+A drawback could be finding OCI registry clients that can deal with these
+formats.
+
```

Comment:
I think the main drawback is that this solutions requires more code to be written into Rancher Manager and potentially extending the airgap mechanism of Rancher manager.

Reply:
expanded in https://github.com/kubewarden/rfc/pull/12/commits/2745bd85485273610b26877d3556def055ed7799.